### PR TITLE
Added resources monitoring

### PR DESCRIPTION
**Pull request by starkmsu**

### CODE REVIEW

File: src/Lykke.Common/ResourcesMonitor.cs [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+            }
+
+            double memoryInMBytes = _process.WorkingSet64 / _1mb;
+            TelemetryHelper.TrackMetric(_ramMetric, memoryInMBytes);
+
+            if (_ramMbThreshold.HasValue && _ramMbThreshold.Value <= memoryInMBytes)
+                _log.WriteMonitor(nameof(ResourcesMonitor), "", $"RAM usage is {memoryInMBytes}");
+
+            _startCpuTime = _process.TotalProcessorTime;
+            _cpuWatch.Restart();
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Common/ResourcesMonitor.cs]: We not takes into account CPU time spent, while code between _cpuWatch.Stop() and _cpuWatch.Restart() is being executed. May be will be more accurate and will make code simpler, if we'll just start _cpuWatch and remember  _startCpuTime once in the ctor and then, in the Execute just will take current _cpuWatch.Ellapsed and difference between _process.TotalProcessorTime and _startCpuTime. _startCpuTime will be not nullable in this case and all CPU time starting from the ResourcesMonitor constructing will be counted.

File: src/Lykke.Common/ResourcesMonitor.cs [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        private readonly double? _cpuThreshold;
+        private readonly int? _ramMbThreshold;
+
+        private TimeSpan? _startCpuTime;
+
+        /// <summary>
+        /// Inits monitoring with ApplicationInsights telemetry submission only.
+        /// </summary>
+        /// <param name="log">ILog implementation</param>
+        public ResourcesMonitor(ILog log)
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Common/ResourcesMonitor.cs]: We really want to expose constructors to the client? May be will be better to make them internal and leave only option to register ResourcesMonitor in the container using extension methods.

File: src/Lykke.Common/TelemetryHelper.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+namespace Common
+{
+    /// <summary>
+    /// Helper class for ApplicationInsights telemetry submission
+    /// </summary>
+    public static class TelemetryHelper
+    {
+        private static readonly TelemetryClient _telemetryClient = new TelemetryClient();
+
+        public static void TrackMetric(string name, double value)
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Common/TelemetryHelper.cs]: Could double be replaced with generic parameter? If not, will be nice to add overloads with common primitive types too

File: src/Lykke.Common/TelemetryHelper.cs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        {
+            _telemetryClient.TrackMetric(name, value);
+        }
+
+        public static IOperationHolder<RequestTelemetry> StartRequestOperation(string name)
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Common/TelemetryHelper.cs]: Can we shortly describe int XML-doc, what the difference between StartRequestOperation and TrackMetric?

File: src/Lykke.Common/TimerPeriod.cs [hunk 5]
```diff
@@ -14,11 +12,11 @@ public interface ITimerCommand
         Task Execute();
     }
 
-    // Таймер, который исполняет метод Execute через определенный интервал после окончания исполнения метода Execute
+    /// <summary>
+    /// Timer that runs Execute method in a loop with a fixed time interval between runs.
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Common/TimerPeriod.cs]: Please, add, thath time spent by Execute method is not included in the period

File: src/Lykke.Common/TelemetryHelper.cs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+using Microsoft.ApplicationInsights.DataContracts;
+using Microsoft.ApplicationInsights.Extensibility;
+using Microsoft.ApplicationInsights.Extensibility.Implementation;
+
+namespace Common
+{
+    /// <summary>
+    /// Helper class for ApplicationInsights telemetry submission
+    /// </summary>
+    public static class TelemetryHelper
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Common/TelemetryHelper.cs]: It's trifle, but I personally hate Helper suffix in class names :) What about just Telemetry?

File: src/Lykke.Common/ResourcesMonitor.cs (hunk 2 above)
Comment (starkmsu) [reply src/Lykke.Common/ResourcesMonitor.cs]: Is there any problem with this?

File: src/Lykke.Common/ResourcesMonitor.cs [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    /// CPU and RAM monitoring background checker based on TimerPeriod with once per minute frequency
+    /// </summary>
+    public class ResourcesMonitor : TimerPeriod
+    {
+        private const double _1mb = 1024 * 1024;
+        private const string _cpuMetric = "Custom CPU";
+        private const string _ramMetric = "Custom RAM";
+
+        private readonly ILog _log;
+        private readonly Process _process = Process.GetCurrentProcess();
```
Comment (MikhailGerasimov) [inline src/Lykke.Common/ResourcesMonitor.cs]: Process is disposable

File: src/Lykke.Common/ResourcesMonitor.cs [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+        /// Inits monitoring that beside ApplicationInsights telemetry submission also logs threshold crossing events on monitor level.
+        /// </summary>
+        /// <param name="log">ILog implementation</param>
+        /// <param name="cpuThreshold">Optional CPU threshold for monitor logging</param>
+        /// <param name="ramMbThreshold">Optional RAM threshold for monitor logging</param>
+        public ResourcesMonitor(ILog log, double? cpuThreshold, int? ramMbThreshold)
+            : base((int)TimeSpan.FromMinutes(1).TotalMilliseconds, log)
+        {
+            _log = log ?? throw new ArgumentNullException(nameof(log));
+            _cpuThreshold = cpuThreshold;
```
Comment (MikhailGerasimov) [inline src/Lykke.Common/ResourcesMonitor.cs]: Check arguments on negative values

File: src/Lykke.Common/TimerPeriod.cs [hunk 9]
```diff
@@ -77,20 +75,20 @@ public virtual Task Execute(CancellationToken cancellation)
...
                     try
                     {
                         await Execute(cancellation);
                     }
                     catch (Exception exception)
                     {
                         await LogFatalErrorAsync(exception);
-                        telemtryOperation.Telemetry.Success = false;
-                        _telemetryClient.TrackException(exception);
+                        TelemetryHelper.MarkFailedOperation(telemtryOperation);
```
Comment (MikhailGerasimov) [inline src/Lykke.Common/TimerPeriod.cs]: OperationCancelled probably is not a fatal error

Comment (KonstantinRyazantsev) [reply src/Lykke.Common/TimerPeriod.cs]: All of exceptions here - is just errors, not fatal erros

File: src/Lykke.Common/ResourcesMonitor.cs (hunk 2 above)
Comment (KonstantinRyazantsev) [reply src/Lykke.Common/ResourcesMonitor.cs]: Expose only necessary API to the client - is good practice. If it's not necessary, it just confuse developer, can lead to incorrect usage and undesirable dependencies, which can obsctruct library modification without braking changes the in future

File: src/Lykke.Common/ResourcesMonitor.cs (hunk 7 above)
Comment (starkmsu) [reply src/Lykke.Common/ResourcesMonitor.cs]: Main process will work as long as the app is running. So no Dispose call is required.

Comment (MikhailGerasimov) [reply src/Lykke.Common/ResourcesMonitor.cs]: Now, when ResourcesMonitor is internal it is true

**Pull request merged**